### ISSUE TITLE
[no-unnecessary-condition] false-positive if boolean changes in callback

### ISSUE THREAD

**Issue by Dionysusnu** [typescript-eslint/typescript-eslint#1984]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": ["error"]
  }
}
```

```TS
declare function callsItLater(callback: (...args: any) => void): void;

async function f() {
	let bool = true;
	callsItLater(() => bool = false);
	while (bool) { // Unnecessary conditional, value is always truthy. (eslint@typescript-eslint/no-unnecessary-condition)
		await Promise.resolve();
	}
}
```

**Expected Result**
No error, because `bool` may get re-assigned to false
**Actual Result**
false-positive error
**Additional Info**
None

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.30.0` |
| `@typescript-eslint/parser`        | `2.30.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.3.0` |
| `pnpm`                              | `4.11.5` |

**Comment by bradzacher** [comment]
![image](https://user-images.githubusercontent.com/7462525/81329226-15c2e180-9053-11ea-8091-ee8acc715add.png)

Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/1459
